pkg/shardddl/optimism/lock.go

```diff
@@ -501,27 +526,123 @@ func (l *Lock) GetVersion(source string, schema string, table string) int64 {
 	return l.versions[source][schema][table]
 }
 
+// IsDroppedColumn checks whether this column is a not fully dropped column for this lock
+func (l *Lock) IsDroppedColumn(info Info, col string) bool {
+	if _, ok := l.columns[col]; !ok {
+		return false
+	}
+	source, upSchema, upTable := info.Source, info.UpSchema, info.UpTable
+	if _, ok := l.columns[col][source]; !ok {
+		return false
+	}
+	if _, ok := l.columns[col][source][upSchema]; !ok {
+		return false
+	}
+	if _, ok := l.columns[col][source][upSchema][upTable]; !ok {
+		return false
+	}
+	return true
+}
+
+// AddDroppedColumn adds a dropped column name in both etcd and lock's column map
+func (l *Lock) AddDroppedColumn(info Info, col string) error {
+	if l.IsDroppedColumn(info, col) {
+		return nil
+	}
+	log.L().Debug("add not fully dropped columns", zap.Stringer("info", info), zap.String("column", col))
+
+	source, upSchema, upTable := info.Source, info.UpSchema, info.UpTable
+	_, _, err := PutDroppedColumn(l.cli, info, col)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := l.columns[col]; !ok {
+		l.columns[col] = make(map[string]map[string]map[string]interface{})
+	}
+	if _, ok := l.columns[col][source]; !ok {
+		l.columns[col][source] = make(map[string]map[string]interface{})
+	}
+	if _, ok := l.columns[col][source][upSchema]; !ok {
+		l.columns[col][source][upSchema] = make(map[string]interface{})
+	}
+	l.columns[col][source][upSchema][upTable] = struct{}{}
+	return nil
+}
+
+// DeleteColumnsByDDLs deletes the dropped columns by DDLs.
+func (l *Lock) DeleteColumnsByDDLs(ddls []string) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	colsToDelete := make([]string, 0, len(ddls))
+	for _, ddl := range ddls {
+		col, err := GetColumnName(l.ID, ddl, ast.AlterTableDropColumn)
+		if err != nil {
+			return err
+		}
+		if len(col) > 0 {
+			colsToDelete = append(colsToDelete, col)
+		}
+	}
+	if len(colsToDelete) > 0 {
+		log.L().Debug("delete not fully dropped columns",
+			zap.String("lockID", l.ID), zap.Strings("columns", colsToDelete))
+
+		_, _, err := DeleteDroppedColumns(l.cli, l.Task, l.DownSchema, l.DownTable, colsToDelete...)
+		if err != nil {
+			return err
+		}
+
+		for _, col := range colsToDelete {
+			delete(l.columns, col)
+		}
+	}
+
+	return nil
+}
+
 // AddDifferentFieldLenColumns checks whether dm adds columns with different field lengths
-func AddDifferentFieldLenColumns(lockID, ddl string, oldJoined, newJoined schemacmp.Table) error {
+func AddDifferentFieldLenColumns(lockID, ddl string, oldJoined, newJoined schemacmp.Table) (string, error) {
+	col, err := GetColumnName(lockID, ddl, ast.AlterTableAddColumns)
+	if err != nil {
+		return col, err
+	}
+	if len(col) > 0 {
+		oldJoinedCols := schemacmp.DecodeColumnFieldTypes(oldJoined)
+		newJoinedCols := schemacmp.DecodeColumnFieldTypes(newJoined)
+		oldCol, ok1 := oldJoinedCols[col]
+		newCol, ok2 := newJoinedCols[col]
+		if ok1 && ok2 {
```

Comment:
we could combine this two level if

Reply:
addressed in https://github.com/pingcap/dm/pull/1510/commits/baee6facb5a910f73e6e9462091240a60d7800ec